packages/boba/contracts/contracts/Teleportation.sol

```
@@ -273,30 +269,40 @@ contract Teleportation is PausableUpgradeable {
-        require(supportedChains[_toChainId], "Target chain is not supported");
+        SupportedToken memory supToken = supportedTokens[_token];
+        require(supToken.supported == true, "Token not supported");
+        require(_amount >= supToken.minDepositAmount, "Deposit amount too small");
+        require(_amount <= supToken.maxDepositAmount, "Deposit amount too big");
+        require(supportedChains[_toChainId], "Target chain not supported");
+        // minimal workaround to keep logic concise
+        require(address(0) != _token || _amount == msg.value, "Native amount invalid");
 
         // check if the total amount transferred is smaller than the maximum amount of tokens can be transferred in 24 hours
         // if it's out of 24 hours, reset the transferred amount to 0 and set the transferTimestampCheckPoint to the current time
-        if (block.timestamp < transferTimestampCheckPoint + 86400) {
-            transferredAmount += msg.value;
-            require(transferredAmount <= maxTransferAmountPerDay, "max amount per day exceeded");
+        if (block.timestamp < supToken.transferTimestampCheckPoint + 86400) {
+            supToken.transferredAmount += _amount;
+            require(supToken.transferredAmount <= supToken.maxTransferAmountPerDay, "max amount per day exceeded");
         } else {
-            transferredAmount = msg.value;
-            require(transferredAmount <= maxTransferAmountPerDay, "max amount per day exceeded");
-            transferTimestampCheckPoint = block.timestamp;
+            supToken.transferredAmount = _amount;
+            require(supToken.transferredAmount <= supToken.maxTransferAmountPerDay, "max amount per day exceeded");
+            supToken.transferTimestampCheckPoint = block.timestamp;
+        }
+
+        supportedTokens[_token] = supToken;
+        if (_token != address(0)) {
+            IERC20(_token).safeTransferFrom(msg.sender, address(this), _amount);
```

with opening this method for all tokens, is there a re-entrancy?

---

- All tokens are whitelisted by us. Only the ZeroAddress for indicating native assets is added by default.
- Other safeTransfer(From) calls are within the onlyOwner modifier, so only to be called by us. 
- for `teleportAsset()`: even tough tokens are added via whitelisting, one shouldn't be able to conduct re-entrancy (and even if somehow possible at least limit it) as 

1. `transferTimestampCheckPoint` is updated before the transfer and would let the reentrancy fail.
2. Further limiting it: `supToken.transferredAmount` will be checked against maxTransferAmountPerDay before doing the transfer.

Happy for feedback

---

makes sense!